rpcs3/Emu/RSX/RSXFIFO.cpp

```diff
@@ -58,16 +66,115 @@ namespace rsx
 			}
 		}
 
-		void FIFO_control::set_get(u32 get, bool check_spin)
+		std::pair<bool, u32> FIFO_control::fetch_u32(u32 addr)
 		{
-			if (check_spin && m_ctrl->get == get)
+			if (addr - m_cache_addr >= m_cache_size)
 			{
-				if (const u32 addr = m_iotable->get_addr(m_memwatch_addr); addr + 1)
+				const u32 put = read_put();
+
+				if (put == addr)
 				{
-					m_memwatch_addr = get;
-					m_memwatch_cmp = vm::read32(addr);
+					return {false, FIFO_EMPTY};
 				}
 
+				m_cache_addr = addr & -128;
+
+				const u32 addr1 = m_iotable->get_addr(m_cache_addr);
+
+				if (addr1 == umax)
+				{
+					m_cache_size = 0;
+					return {false, FIFO_ERROR};
+				}
+
+				if (!g_cfg.core.rsx_accurate_fifo_access)
+				{
+					std::memcpy(&m_cache, vm::base(addr1 + (addr & 0x60)), 32);
+					m_cache_addr = addr & -32;
+					m_cache_size = 32;
+					return fetch_u32(addr);
+				}
+
+				m_cache_size = std::min<u32>((put | 0x7f) - m_cache_addr, u32{sizeof(m_cache)} - 1) + 1;
+
+				if (0x100000 - (m_cache_addr % 0x100000) < m_cache_size)
+				{
+					// Check if memory layout changes in the next 1MB page boundary
+					if ((addr1 >> 20) + 1 != (m_iotable->get_addr(m_cache_addr + 0x100000) >> 20))
+					{
+						m_cache_size = 0x100000 - (m_cache_addr % 0x100000);
+					}
+				}
+
+				// Make mask of cache lines to fetch
+				u8 to_fetch = static_cast<u8>((1u << (m_cache_size / 128)) - 1);
+
+				if (addr < put && put < m_cache_addr + m_cache_size)
+				{
+					// Adjust to knownly-prepared FIFO buffer bounds 
+					m_cache_size = put - m_cache_addr;
+				}
+
+				rsx::reservation_lock<true, 1> rsx_lock(addr1, m_cache_size, true);
+
+				const auto src = vm::_ptr<spu_rdata_t>(addr1);
+
+				// Find the next set bit after every iteration
+				for (u32 i = 0, wait = 0;; i = (std::countr_zero<u32>(utils::rol8(to_fetch, 0 - i - 1)) + i + 1) % 8)
+				{
+					// If a reservation is being updated, try to load another
+					const auto& res = vm::reservation_acquire(addr1 + i * 128);
+					const u64 time0 = res;
+
+					if (!(time0 & 127))
+					{
+						mov_rdata(m_cache[i], src[i]);
+
+						if (time0 == res && cmp_rdata(m_cache[i], src[i]))
+						{
+							// The fetch of the cache line content has been successful, unset its bit
+							to_fetch &= ~(1u << i);
+
+							if (!to_fetch)
+							{
+								break;
+							}
+
+							continue;
+						}
+					}
+
+					busy_wait(200);
+
+					if (++wait == 5000u)
+					{
+						wait = 0;
+
+						const auto rsx = get_current_renderer();
+
+						if (rsx->is_stopped())
+						{
+							return {};
+						}
+
+						rsx->cpu_wait({});
+					}
+				}
+			}
+
+			be_t<u32> ret;
+			std::memcpy(&ret, reinterpret_cast<const u8*>(&m_cache) + (addr - m_cache_addr), sizeof(u32));
+			return {true, ret};
+		}
+
+		void FIFO_control::set_get(u32 get, u32 spin_cmd)
+		{
+			m_cache_size = 0;
```

Comment:
It is important to get stats about how much data we're throwing away. Call invalidate_cache here and we can have stat counters there. The cache load is going to be a new area for further tuning is my suspicion.
There is a frame_stats object in rsx::thread, but I prefer if FIFO_control has its own stats structure with FIFO reads vs misses (discards). I can handle adding them to the debug overlay later, they just need to be made available.